Deployment refactor (#683)
scalacenter/scastie: CONTRIBUTING.md

# How to deploy

## Preparing SBT environment
The whole deployment process is done semi-automatically via SBT task.

```
ssh scastie@alaska.epfl.ch
ssh scastie@scastie.scala-lang.org
ssh scastie@scastie-sbt.scala-lang.org
docker login
cd ~/scastie && git pull && ~/nix-user-chroot-bin-1.2.2-x86_64-unknown-linux-musl ~/.nix bash -l
nix-shell -v
sbt
```

Scastie offers multiple deployment configurations with following instructions for each of them.

> If you didn't make any changes to the deployment task should finish with success.

Reviewer: ```suggestion
If you didn't make any changes to the deployment task, it should finish with success.
```

Otherwise, the deployment will fail and inform you that remote deployment scripts are not up-to-date with their
local counterparts. In such situation you should follow the instructions from the SBT console.

Every IP which is put into the configuration should be double-checked. If you have any doubts if they are correct, please
contact one of the maintainers.

## Production

> Scastie production environment is used for internal deployment. It's configuration is present at `./deployment/production.conf`.

Reviewer: ```suggestion
Scastie production environment is used for internal deployment. Its configuration is present at `./deployment/production.conf`.
```

> The production deployment is start with `deploy` command.

Reviewer: ```suggestion
Start the production deployment with the `deploy` sbt command.
```

## Staging

> Scastie also has staging environment. The deployment can be done by running `deployStaging` task.

Reviewer: ```suggestion
Scastie also has a staging environment. The deployment can be done by running the task `deployStaging`.
```

It will do normal deployment, but with Staging environment configuration file located at: `./deployment/staging.conf`

## Dry run

Scastie deployment process generates shell scripts responsible for proper remote deployment.
If in any case you want to check the scripts without running them, it can be done by running
`deployDryRun` SBT task.

## Reverting the deployment

> Deployment process is sequentiall, so if the process fails on any step, it will be stopped so only one part of the system

Reviewer: ```suggestion
The deployment process is sequential, so if the process fails on any step, it will be stopped so only one part of the system
```

> will be affected. You will need to manually start failed step.

Reviewer: ```suggestion
will be affected. You will need to manually restart the failed step.
```
